File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
    vector<bool> visited;
    vector<bool> detect;
public:
    bool canFinish(int n, vector<vector<int>>& p) {
        vector<vector<int>> adj(n);
        detect.resize(n, false);
        visited.resize(n, false);
        for (auto const pairCourses : p) {
            adj[pairCourses[0]].push_back(pairCourses[1]);
        }

        for (int i = 0; i < n; ++i) {
            if (!visited[i]) {
                if (dfs(i, adj)) return false; // detected cycle
            }
        }
        return true;
    }

    bool dfs(int i, const vector<vector<int>>& adj) {
        visited[i] = true;
        detect[i] = true;

        for (int neighbor : adj[i]) {
            if (!visited[neighbor]) {
                if (dfs(neighbor, adj)) return true;
            }
            else if (detect[neighbor]) {
                return true; // detected cycle
            }
        }

        detect[i] = false; // Reset detect status after visiting all neighbors
        return false;
    }
};

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, n: int, p: List[List[int]]) -> bool:
        
        ans = False
        adj = [[] for _ in range(n)]
        visited = [False] * n
        detect = [False] * n        
        for i in range(len(p)):
            adj[p[i][0]].append(p[i][1])
            
        def dfs(i):
            visited[i] = True
            detect[i] = True
            neighbors = adj[i]
            for neighbor in neighbors:
                if detect[neighbor]:
                    return True
                if not visited[neighbor] and  dfs(neighbor):
                    return True
            detect[i] = False
            return False
        for i in range(n):
            if not visited[i] and dfs(i):
                return False
            
        return True